Fix logging RPC and -debugexclude with 0/none values, add test coverage, improve docs

jonatack opened the pull request:
The logging RPC does not work as intended and as documented in its help when `0` or `none` are passed.

Per the RPC logging help:
```
$ ./src/bitcoin-cli help logging

In addition, the following are available as category names with special meanings:
  - "all",  "1" : represent all logging categories.
  - "none", "0" : even if other logging categories are specified, ignore all of them.
```
    
The behavior and documentation were added in #11191, but over time and extended refactoring, the behavior no longer works in master and versions 22/23/24 (the last supported versions currently) -- passing `0`/`none` has no effect.  As there was no test coverage, the regressions went uncaught.  In v24, `none` became unrecognized:

```
$ ./src/bitcoin-cli logging '["none"]'
error code: -8
error message:
unknown logging category none
```
    
During the same time period, passing `1` and `all` has been operational and documented. 

Solution: detect `none`/`0` values and add test coverage in any case, and either:
    
- leave the functionality out, raise with an error message if the values are passed, and update the RPC help documentation, or

- fix the behavior by returning early.

Both solutions involve essentially the same code size and complexity.  Given that `all`/`1` has been operational, and that `none`/`0` has been documented and appears per the code of the original PR to have been operational, it seems preferable for consistency to support `none`/`0` in symmetry with `all`/`1` and as ACKed, intended, and documented in #11191.

Done by this pull:    
- add missing RPC logging logic and test coverage
- update the `-debug` and `-debugexclude` config options to use the same code for consistency in behavior between them, and for consistency with the logging RPC that provides the same behavior for both `-include` and `-exclude`
- improve the user-facing documentation and examples; can be tested with `./src/bitcoind -h | grep -A12 "\-debug=<\|debugexclude" && ./src/bitcoin-cli help logging`

If it is decided to backport a fix, commit `Fix RPC logging behavior when "0" and "none" values are passed` could probably suffice.

jonatack committed:
Functional test coverage for RPC logging include/exclude with all/none

jonatack committed:
Improve RPC logging documentation and examples

and make the logging example in the developer notes zsh-compatible
(for instance, zsh is currently the default shell in macOS).

The RPC logging help can be tested with:

./src/bitcoin-cli help logging

fanquake commented:
>  or none are passed, including not recognizing none:

Any idea when/how this was broken?

fanquake commented:
Probably in f1379aeca9d3a8c4d3528de4d0af8298cb42fee4 from #25614:
> replace the unused BCLog::Level:None string "none" with an empty string
  as the case will never be hit

So only broken for 24.x.

jonatack committed:
Fix RPC logging behavior when "0" and "none" values are passed

Per the RPC logging help:

```
In addition, the following are available as category names with special meanings:
  - "all",  "1" : represent all logging categories.
  - "none", "0" : even if other logging categories are specified, ignore all of them.
```

The behavior and documentation were added in #11191, but over time and extended
refactoring, the behavior is not longer working in Bitcoin Core versions
22/23/24 (the last supported versions currently) -- passing `0` or `none` has no
effect.  As there was no test coverage, the regressions went uncaught.  In v24,
`none` became unrecognized.

During the same time period, passing `1` and `all` has been operational and documented.

We therefore need to detect none/0 values and add test coverage in any case, and either:

- leave the functionality out, raise with an error message if none/0 are passed,
  and update the RPC help documentation, or

- fix the behavior by returning early.

Both solutions involve essentially the same code size and complexity.  Given
that all/1 has been operational, and that none/0 has been documented and appears
to have been operational previously, it seems preferable for consistency to
support none/0 in symmetry with all/1, and as ACKed, intended, and documented
since #11191.

---

This commit therefore adds:

- the missing logic to make the functionality work

- a missing logging.h include header

- the relevant info in the logging RPC help doc

jonatack committed:
Use same code/behavior for -debug and -debugexclude config options

for consistency between them and for consistency with the logging RPC that uses
the same logic for both -include and -exclude.

Co-authored-by: Anthony Towns <aj@erisian.com.au>
Co-authored-by: Matthew Zipkin <pinheadmz@gmail.com>

jonatack committed:
Functional test coverage for -debug/-debugexclude config options

and convert some comments to logging.

fanquake commented:
> The behavior was intended to be added in https://github.com/bitcoin/bitcoin/pull/11191.

It's still not clear if this was broken from the time it was merged, or at some other later point. This also does not work with 22.x, and any earlier versions are EOL.

If it's been broken forever (or at least is in all currently maintained releases), and no ones even noticed, we could take advantage of that, to remove some of the complexity here; do we definitely need multiple different ways of achieving the same logging toggling?

jonatack committed:
Update -debug and -debugexclude help documentation

Can be tested with:

./src/bitcoind -h | grep -A12 "\-debug=<\|debugexclude"

jonatack committed:
log: remove unused BCLog::NONE enums in logging.cpp code

These cases should never be hit and are not needed; the 0/none logging category
logic is handled only by the callers of IsNoneCategory() at these times:

1) during bitcoind startup init

2) when the `logging` RPC is invoked

pinheadmz commented on src/init/common.cpp:
```diff
@@ -78,27 +79,24 @@ void SetLoggingLevel(const ArgsManager& args)
+    const std::vector<std::string>& categories{args.GetArgs(opt)};
+    if (std::any_of(categories.cbegin(), categories.cend(), [](const auto& c) { return LogInstance().IsNoneCategory(c); })) return;
+    for (const auto& c : categories) {
+        const bool success{opt == DEBUG_LOG_OPTS[0] ? LogInstance().EnableCategory(c) : LogInstance().DisableCategory(c)};
```
`DEBUG_LOG_OPTS[0] `

Feels like this is undoing whatever convenience/readability you set up by defining `DEBUG_LOG_OPTS` in the first place. Do you think more debug args will be added in the future? Or can we just use

`args.IsArgSet("-debug") || args.IsArgSet("-debugexclude")` ... etc

pinheadmz replied:
Well I meant something like a switch statement with `"-debug"` and `"-debugexclude"` literal strings as cases. I think that would make it more readable and if anything is ever added to `DEBUG_LOG_OPTS[]` it won't necessarily break. But that might just be too long or out of style

pinheadmz replied:
Yeah 😬 waddaya think?

pinheadmz commented on src/init/common.cpp:
```diff
@@ -26,10 +26,10 @@ namespace init {
-        "If <category> is not supplied or if <category> = 1, output all debug and trace logging. <category> can be: " + LogInstance().LogCategoriesString() + ". This option can be specified multiple times to output multiple categories.",
+    argsman.AddArg("-debug=<category>", "Output debug and trace logging (default: 0). "
+        "If <category> is omitted, or is 1 or all, output all debug and trace logging. If <category> is 0 or none, any other categories passed are ignored. Other valid values for <category> are: " + LogInstance().LogCategoriesString() + ". This option can be specified multiple times to output multiple categories. See also the -debugexclude configuration option that takes priority over -debug.",
         ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
-    argsman.AddArg("-debugexclude=<category>", "Exclude debug and trace logging for a category. Can be used in conjunction with -debug=1 to output debug and trace logging for all categories except the specified category. This option can be specified multiple times to exclude multiple categories.", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-debugexclude=<category>", "Exclude debug and trace logging for a category (default: 1). Takes priority over -debug and can be used in conjunction with -debug=1 to output debug and trace logging for all categories except the specified category. If <category> is 1 or all, exclude all debug and trace logging. If <category> is 0 or none, any other categories passed are ignored. O ...
```
Recommend putting "all" and "none" in quotes to be super-duper explicit

pinheadmz commented on src/rpc/node.cpp:
```diff
@@ -238,22 +238,31 @@ static RPCHelpMan logging()
                 RPCExamples{
-                    HelpExampleCli("logging", "\"[\\\"all\\\"]\" \"[\\\"http\\\"]\"")
-            + HelpExampleRpc("logging", "[\"all\"], [\"libevent\"]")
+                   "\nTurn on tor and i2p logging categories\n"
+                   + HelpExampleCli("logging", "'[\"tor\", \"i2p\"]'")
+                   + HelpExampleCli("-named logging", "include='[\"tor\", \"i2p\"]'")
+                   + "\nTurn on all logging categories, except http\n"
+                   + HelpExampleCli("logging", "'[\"all\"]' '[\"http\"]'")
+                   + HelpExampleCli("-named logging", "include='[\"all\"]' exclude='[\"http\"]'")
+                   + "\nTurn off all logging categories\n"
+                   + HelpExampleCli("logging", "'[]' '[\"all\"]'")
+                   + HelpExampleCli("-named logging", "exclude='[\"all\"]'")
+                   + "\nTurn on all logging categories, except libevent\n"
+                   + HelpExampleRpc("logging", "[\"all\"], [\"libevent\"]")
```
GREAT

pinheadmz reviewed (commented):
concept ACK
github code review, will fine-tooth comb review later this week.
Played with build locally and tried out examples from new help message. A few questions...

jonatack committed:
Functional test that RPC logging does not return all/none categories

jonatack committed:
Unit tests for LogPrintfCategory and LogPrintLevel with BCLog::ALL/NONE

to specify expected current behavior and check future changes for regressions.

pinheadmz commented on src/logging.cpp:
```diff
@@ -225,7 +224,7 @@ std::string LogCategoryToStr(BCLog::LogFlags category)
     switch (category) {
     case BCLog::LogFlags::NONE:
-        return "";
+        return "0";
```
Ah I see, these cases are essentially handled by `IsNoneCategory()` 👍

pinheadmz reviewed (approved):
ACK e9f6fd0e028e6a5669926926792c70c31a26c403

reviewed code, built and ran tests locally. played with feature in regtest, tried to break it, couldn't break it.

just one non-busting question below

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK e9f6fd0e028e6a5669926926792c70c31a26c403
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmQ29zkACgkQ5+KYS2KJ
yTqh4hAAwBBMTmfeyYhdSTLpO2hma6GolZdUdvzOYdMhADsucUvihZIic8o70MZU
hzoqnYsGdTvw31MxoBDWre10y9jBvXfK0vtrkC8A7nzMNBHVaVTqFfsVayUQ4TMm
6PPrKJm1TbdAUc5/aiv6B3FEiCdUx2hlugWyFnzxLt3sPS8pnPMQ8GMdyxo7+0Mh
OYsRLxCvwnMFHMGh/mA3do7Znj28f3QM4xKGYV7OA/GeghqNl75Oo3+7pGB/Rq5D
JXfqjLN4/pAvZ78dr5k6O8Xgz/tlsCtPs5gJo+aaVwPv0my5fpCO3GbegD76ybuK
3teZPSCufYe7XBNOQVnkXB4EtnhwrXPEXz/xAAOssBKhQrfk7TV4sovi4pRrNxec
rAVO3Fc9H0weHywxi1DO56/0F9oLNVBWJMLmV1t9/FF/lrIHvWDTwvywKg1N/mpQ
MY4NKmBLVD0kLdczmEhdpPua2400WwYWm/ypPXjOJhGuS7pmJW2WNOXv06NNXaxm
h/4byxYhgB6AkffZ5IQM5+6gEXixWbFDmV/eEciydB0VnNv1+nJctZ1QxyVUnaZ9
ZfBRm9e99Q9JHtOkvF8UF3X5bLP9bCqOhCThqjWsxy6shzpkkkhBE0O+HDkJEHsz
cyKvlcNo4a2OX0EkITmX195G3MYqFK0RATitcM5e3SZEA7wkdVg=
=DtQx
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

pinheadmz commented on src/init/common.cpp:
```diff
@@ -78,27 +77,26 @@ void SetLoggingLevel(const ArgsManager& args)
-        if (!LogInstance().DisableCategory(cat)) {
-            InitWarning(strprintf(_("Unsupported logging category %s=%s."), "-debugexclude", cat));
-        }
+void SetLoggingCategories(const ArgsManager& args)
+{
+    for (const std::string& opt : DEBUG_LOG_OPTS) {
+        EnableOrDisableLogCategories(args, opt);
```
Cool, just to be clear I meant combining this `EnableOrDisableLogCategories()` into ` SetLoggingCategories()`  and then you only have one named `EnableOrDisableLogCategories()` in the code over in node.cpp

> That, along with how extracting that function simplifies SetLoggingCategories to highlight that both options are being plugged into the same code, were the motivations.

okie doke 👍

pinheadmz reviewed (approved):
ACK 5991eddb98ce59b883ae695752d4e90f32d43960
<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 5991eddb98ce59b883ae695752d4e90f32d43960
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmQ4FuQACgkQ5+KYS2KJ
yTqVKQ//TIOVSamur/OBGsHB1lgx9C3wyqQFSg+RzImVrDvquEhTDdwfjWI56DzL
sYWxaSDjHLP+4jYK7x3O4QoZJW1iJDlz56PhakxtDSzWMIr0WIfHGpTV6oH1p4Uh
Kaz57R6Y71G4ADMqeR48MqLK/B7vMuEKhzUUmGk14O/8bPUHk7HJE70b5F3zOTb9
EBu/NtFd3D5rdRCSTMbmkvkuZ8ezAzoROTT+X08ge1Rf2RwVHeMklY74apx12+qT
n4xklbM1YZ/6HN40uCxGXnhgYtqhwAk8I94Zz5J8lJTh7UbXDrYcWiyGd3x6Ph7Y
6ZmSP7cgCXTAHx92Hb6QHysInJZVcd/SP11PhdWkgSAvLpKHcXiZiPOMUKz6p4Ih
O3i/BvsaGR0HkPxCQ0x9XDoABHFGNT3YT64dYiCxbwBb9yPiNzZYTOexH8iZOIQp
iftnh/d+iZSIiPZP1uVctfrUsr9U+ECIV0hFOheC7yDygxYI3QIyzFYj2kzxfm0c
YwBAgCSnl0uxGHryFtya3fr5Y0zFk+wlhhm04nLM5VVbxq3kQmfg8+CKPJ8MayCh
LRc89eL53ijefy5pmCXEZvMLJIX9bcdLVTRJPe4JS5lRh+fmSP2rEyPffymnXL1P
51es/2so7kOxu4chEr178/+GY4uZjIPxd1MOMxl11QE7ug8COXQ=
=sU3Z
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

pinheadmz commented on src/init/common.cpp:
```diff
@@ -80,26 +79,31 @@ util::Result<void> SetLoggingLevel(const ArgsManager& args)
+    NONFATAL_UNREACHABLE();
+}
+
+// Maintain similar logic in both of the EnableOrDisableLogCategories() functions in the codebase.
```
I still don't totally understand why you need to split this up into three functions, one of which is a kinda-duplicate The comment makes sense but I'm not familiar enough with the general style of the project to determine if having two functions with the same name doing basically the same thing is acceptable.

pinheadmz replied:
OK cool thanks for explaining

pinheadmz reviewed (approved):
re-ACK 630b38e92b4e40d49d69970b6c5bc02aa39ba210

Confirmed refactor since last ACK, making use of `util:Result`

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 630b38e92b4e40d49d69970b6c5bc02aa39ba210
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmSUqggACgkQ5+KYS2KJ
yTqixA/+NOb+9zG4txOdmcIJWSVwPIu7nU59E9EfRvutPPMOw12R0tqm5+B1TvYQ
LxZCFVTck5fe+dAjy0v0wEimrfyY7fJzfxp+HC+ksNWkh88xrgfVFPIzSLwmHx4Z
Ake54SuSEICD7FUJ8G43Q5uNkgG+/rzQ73iLdiY9fdmex6y2kW5yXoPlwhczjRJo
lGV7JXVC0yZjwuey9LhQdCHpyr0rc9rUZ3MD3Wfc4fH+NH/20VkOQimA1ux1KSe4
wY0ezbMW0AyeH3WAnnk+c3G8+Y+k88s4f6J/uxfKuyFz52Cz1vfyK/vudabqTvxL
AyZ/P++n1gqBZkyOmzHhKYxv0jtFBHb5OtIj25tZoP33/AlL3iy8ZjD6aUN3tYfU
8dzRiUD7vTYauKnd6QKyGJEGKir8D24UVE34PUM7o1l2GtwK386s3Lr4CfRdVYyP
P0BkL638CgkMzyF5HhQVLoxr/jdpOQdvzd3KY6yzenMdBH8aD5TydwBswvGb4EGd
hvn7txdklo+ZkJLc4A4b7zNjB32NWx2JWUqTr8qV/oKgUPrpo9q2SPB3kqOB9zhV
L1b7VJxVE9xreteU+zOgVVQ9PRtgzXSRJUnC+nCgCZrJm/MMy9iqN54aLFdMmDCf
Xy3kkky33ElyubH7r6NHYARqXoDoxfmMJm3J5EQ72SSDI+j8KX8=
=cmua
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

ajtowns commented on src/init/common.cpp:
```diff
@@ -80,26 +79,31 @@ util::Result<void> SetLoggingLevel(const ArgsManager& args)
+util::Result<void> SetLoggingCategories(const ArgsManager& args)
+{
+    // debugexclude settings take priority over debug ones, so run debugexclude last
+    for (const std::string& opt : {"-debug", "-debugexclude"}) {
```
This seems a bit clunky, how about:

```c++
// Maintain similar logic in both of the EnableOrDisableLogCategories() functions in the codebase.
static void EnableOrDisableLogCategories(util::Result<void>& result, const ArgsManager& args, const std::string& opt, bool enable)
{
    if (!result) return;
    if (!args.IsArgSet(opt)) return;
    const std::vector<std::string>& categories{args.GetArgs(opt)};
    if (std::any_of(categories.cbegin(), categories.cend(), [](const auto& c) { return LogInstance().IsNoneCategory(c); })) return;
    for (const auto& category : categories) {
        bool success = (enable ? LogInstance().EnableCategory(category) : LogInstance().DisableCategory(category));
        if (!success) {
            result = util::Error{strprintf(_("Unsupported logging category %s=%s."), opt, category)};
            return;
        }
    }
    return;
}

util::Result<void> SetLoggingCategories(const ArgsManager& args)
{
    util::Result<void> result;

    // debugexclude settings take priority over debug ones, so run debugexclude last
    EnableOrDisableLogCategories(result, args, "-debug", /*enable=*/true);
    EnableOrDisableLogCategories(result, args, "-debugexclude", /*enable=*/false);
    if (!result) return util::Error{util::ErrorString(result)};
    return {};
}
```

(or just `return result;` at the end -- not sure what the point of `return Error(ErrorString(result))` is?)

pinheadmz replied:
Great suggestion AJ this are looks a lot better now

ajtowns reviewed (commented):
utACK 630b38e92b4e40d49d69970b6c5bc02aa39ba210

luke-jr commented on test/functional/feature_logging.py:
```diff
@@ -14,52 +14,76 @@ class LoggingTest(BitcoinTestFramework):
         assert os.path.isfile(self.relative_log_path("foo.log"))
+        # In addition to net and tor, leveldb/libevent/rand are excluded by the test framework.
+        for category in ["leveldb", "libevent", "net", "rand", "tor"]:
+            assert not node.logging()[category]
```
Kinda ugly to call `node.logging()` repeatedly

luke-jr commented on test/functional/feature_logging.py:
```diff
@@ -14,52 +14,76 @@ class LoggingTest(BitcoinTestFramework):
+            "-debugexclude=qt",
+        ])
         assert os.path.isfile(tempname)
+        assert all(v == True for v in node.logging().values())  # -debugexclude=[none,qt] does not exclude qt
```
Why aren't libevent/leveldb being excluded here?

pinheadmz reviewed (approved):
re-ACK 93b387fe70174d174b840016da8531ef65ec313e

Confirmed changes were compiler warning fix, better test comment, and very nice clean-up in common.cpp. Built branch and ran tests locally.

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 93b387fe70174d174b840016da8531ef65ec313e
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmS5WPgACgkQ5+KYS2KJ
yTp7mxAAwSLI2Fqz8+sRsHF7c/wzERlqZm7Ma8tg9ob0yyvD7Q0t89rCtcDSKL9c
dO2l5DcHotexApg26sp3gfRlQCB7Pv24aCyHeCYl336AcXL57YHbbDu9f/Ct3PRE
+3QoK9b4BiJ2D3db77gJGgmq9QhrnCmCk41bRA5VIyAi3scy4y9lzisRjRswIDji
PnbSJR2qZVO9iDux5Ia7KSHsLxjjC7KhhbhQdDq0LzyK0tQvnyC6+2S8yZVDhxv4
+XyUgsU+YBv49eYPrHid6MVjH58QBBjwBfLGccnqI/ZPb5nogmfFjokj4W8/QRiy
W5nMrCPZYXwRspSn+XbX3Ivzez9mw51Zj3CvQ0tABgMFOTx1ZUESEYwk1SgxjTyV
p2jFMinQVBYeH77IygL2AWeHoha1FCr4zlaLRDBd08MYVccMK+xEL4XGbsNOYWen
mQlBbpPB49YcB2Vs+ZAlHVX4Bp5TKpfVSPEttAyJrn2nXXGPde2gbB3gOIabLK8U
1fAEeG8JZW5Qs4Hl46B51SdpdCe64WkY2lKTWwNZjpmL5YM0ufbPfPscfnpd3Myg
bJfj+6P1N9UhizVYN+Bzm/zDKS/S6zv4DvV8FF13HhGVtlQ+BhGmRa1qr6xSilu8
hxJTLm6bs+2a2/dFQVuhgMMz3Uwt/6tZsIkyX5AgmouDGSkGbTQ=
=kvhB
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

fanquake commented on src/logging.cpp:
```diff
@@ -288,6 +284,8 @@ std::string LogCategoryToStr(BCLog::LogFlags category)
         return "scan";
     case BCLog::LogFlags::ALL:
         return "all";
+    case BCLog::LogFlags::NONE: // this case is normally never hit
```
I think we should avoid adding comments like this, because they don't really provide any value to the developer. If anything, a comment explaining what the "not normal" case is, and/or why that's particularly relevant, would be more useful, because that's the first question I'd have after reading this comment. Also unsure why this code has been re-ordered.

ajtowns commented on src/logging.cpp:
```diff
@@ -288,7 +285,9 @@ std::string LogCategoryToStr(BCLog::LogFlags category)
         return "all";
-    }
+    case BCLog::LogFlags::NONE:
+        Assume(false);
```
What's the point of an `Assume(false)` that just falls through to an `assert(false)` ?

ajtowns reviewed (commented):
ACK 05143b2b94cc1d62f27fc73212aa59f9d471f2b8

ajtowns commented on src/logging.cpp:
```diff
@@ -288,7 +284,9 @@ std::string LogCategoryToStr(BCLog::LogFlags category)
         return "all";
-    }
+    case BCLog::LogFlags::NONE:
+        assert(false);
```
You'd call `Log(std::nullopt, BCLog::Level::Debug)` in that case (or have `constexpr auto NONE = std::nullopt<LogFlags>` and call `Log(BCLog::NONE, BCLog::Level::Debug)` as normal). If `NONE` isn't safe for some functions (eg here where it results in an `assert(false)`), we should just express that in the type system.

ajtowns commented:
reACK 94b059ffe403a948957152b96f24fba4d9ee31f0

pinheadmz reviewed (approved):
re-ACK 94b059ffe403a948957152b96f24fba4d9ee31f0

confirmed with range diff 93b387fe70 to 94b059ffe4
Changes since last review address AJ's comments, minimal diff
Built and tested again locally

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 94b059ffe403a948957152b96f24fba4d9ee31f0
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmT4nZIACgkQ5+KYS2KJ
yTpNDQ//Qlf3FswTPrPl8KG9iZAZvzA0CVvIeBRdFZWgtDJ1cGnIllxEsnVfE7R9
J6EXUEdZ+XCvzfVxLaLR9W23dWtRttNjC+cx/i7GexIYlAQfI4FGECuMitXXEYsR
pYujKEQ/6SETiI8TBJC3OqiWhBodDzzfzdevhiKKbX2dKY9c22OqTSP9TI6XJQfw
R8iqZuj4Tps4jUrNYoWQmo8C71KReVnCdDi3jeKKKXeGFfkb/hJNDqRRqtyLvqd2
Ih/N67BbjtHjtJ2U16x2qFN5xyyBer/nXRirM9GKTKrt6ZpDhY6k2k3h5pVOfKGg
JCUyKLh7N1sWplUbHVQQ+1SHSM9deuwHu/Ld9iuRcBydz4EM9J4oxskVE68CZclT
zynxdsAxoWvxuADMsKj94WY2zsJTr/mnJaq+d6prXTYTkdpxCi9BSno5GxJrIfRX
H7gwqAVxsx8npJbsUmkawB8NBI9xeeYiam1jhipFsWLyovPEAw5APao3/kuVNu58
sMz9ui0dUohtfbPsZoUbimSNofSc5anELjAxT3ekQFImCxHaZSf6/LQ+muNG/qRg
aGrK4fNXsRQX39BIv4E0fYTMm8Ud+r6LuRDTs+z1OEOt+VKVcGs4amQ4jMdtZios
nPeS3TZzKwWVYsjQCZ2LQyICDPkAIFxIRHLWtubrh/WYPM+Leik=
=jmH2
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

ajtowns commented:
> > The behavior was intended to be added in #11191.
> 
> It's still not clear if this was broken from the time it was merged, or at some other later point. This also does not work with 22.x, and any earlier versions are EOL.
> 
> If it's been broken forever (or at least is in all currently maintained releases), and no ones even noticed, we could take advantage of that, to remove some of the complexity here; do we definitely need multiple different ways of achieving the same logging toggling?

Has this comment being replied to anywhere already?

It does seem like it would be pretty easy to remove this behaviour -- removing "none" would mean that anyone who specified it would get an error on startup, which is a good clue to fix their config. And you get the same "actually, for this run I want to disable debugging stuff" by adding `-nodebug` at the end (eg, on the cli) or by specifying "-debugexclude=all". For the logging rpc, just replace the argument with `[]` instead of adding none seems reasonable.

ajtowns commented:
> Exploring this, it doesn't look like there would be much code simplification gained by dropping `none` for `0` only, which would no longer be in symmetry with `all/1` that have been operational for a long time.

Why would you keep the `0` option? Just drop the concept entirely: https://github.com/bitcoin/bitcoin/compare/master...ajtowns:bitcoin:202309-debug-no-none That special case was added well prior to `-debugexclude` (namely #3067 vs #1448

ajtowns commented:
> > Why would you keep the `0` option? Just drop the concept entirely: [master...ajtowns:bitcoin:202309-debug-no-none](https://github.com/bitcoin/bitcoin/compare/master...ajtowns:bitcoin:202309-debug-no-none) That special case was added well prior to `-debugexclude` (namely #3067 vs #10123)

> I think that question is addressed in the pull description.

I would have thought it was obvious that I don't see an answer to that question in the pull description, so that's an impressively unconstructive response. In the PR description, you claim:
 
> * leave the functionality out, raise with an error message if the values are passed, and update the RPC help documentation, or
> * fix the behavior by returning early.
>
> Both solutions involve essentially the same code size and complexity.

But that's trivially not true: removing the code obviously results in a lower code size, and it also reduces the number of special cases that have to be documented and understood by users.

No longer convinced this makes sense.

ajtowns commented:
> ```
> Both solutions involve essentially the same code size and complexity.  Given that `all`/`1` has been
> operational, and that `none`/`0` has been documented and appears per the code of the original PR
> to have been operational, it seems preferable for consistency to support `none`/`0` in symmetry with
> `all`/`1` and as ACKed, intended, and documented in #11191.
> ```
> 
> > But that's trivially not true: removing the code obviously results in a lower code size
> 
> No, not if we check for the values and raise with an error message if the values are passed, with relevant test coverage.

We already check for invalid categories and issue an error message for those. Doing the same thing for none/0 doesn't require extra code or extra test coverage.

>  This ought to be done given that the functionality has been documented in the RPC help for around 7 years now, and perhaps also as it's not clear at what point it broke.

FWIW, it broke in #12954 ("This is purely a refactor with no behavior changes." :roll_eyes:) -- previously `getCategoryMask` evaluated all the strings, returning null if "none" was passed in and erroring out if any were invalid; after that PR `EnableOrDisableLogCategories` dropped the special handling of "none" and enabled/disabled each category as it went, which means `logging ["net", "nosuchcategory"]` changed its behaviour from just giving an error, to first enabling net logging, then giving an error.

#11191 was merged 2017-11-30 and #12954 was merged 2018-05-01 so this behaviour only worked for five months if you were running master, and only for the 0.16.x release. 0.16.0 was also the first release that considered `logging` a public api (#11626).
 
(I have an alias `git-history` which runs `git log --ignore-cr-at-eol -M -C -L${2},${3}:${1}`; so `git blame rpc/node.cpp` to get the line numbers of the function, then `git-history rpc/node.cpp 187 204` makes it easy to trace things through. At least until the refactorings get too intense. AAAA+++ excellent alias, highly recommended)

> That is what I meant by `"Solution: detect none/0 values and add test coverage in any case"`, for either way requires code, test coverage, and documentation changes. As I wrote, I prefer the route of fixing what was ACKed, intended, and documented in #11191.

Why? The people who ACKed an old PR aren't gods, and we don't have to take their opinions at the time as gospel. As far as I can see, the other ways we (now) have of getting this behaviour are strictly superior.

The reasons to have that behaviour originally was that "-debug" was originally a bool flag that got upgraded to support multiple distinct categories, and the arg parser at the time wasn't as clever/complicated as it is today. The reason given in 11191 for doing anything was `"0","1" are allowed in both array of <include> and <exclude>. "0" is ignored and "1" is treated same as "all". It is confusing, so forbid them.` so "forbid them" for none/0 still seems pretty reasonable.

> This pull improves a lot of things here. It not only fixes broken functionality while using lesser or equal lines of code (apart from test coverage and added docs), it also adds test coverage that has been missing for a long time and improves the documentation. It would be nice to see this finally move forward, and I appreciate your reviews of it!

These commits have a bunch of extra complexity compared to dropping support for this behaviour:

 * Fix RPC logging behaviour -- bunch of extra code, with a requirement to keep it in sync with code in init, vs dropping the documentation
 * Update -debug help docs -- bunch of extra text to read documenting a special case that most people won't want
 * Functional test coverage for RPC -- the 0/none loop can be omitted entirely (unknown category just means previously listed categories get actioned)

Other commits seem mostly fine with 0/none special casing dropped:

 * Unit tests for LogPrintfCategory -- fine
 * Functional test does not return all/none -- fine
 * Improve RPC logging doc and examples -- fine
 * Functional test coverage for -debug -- replace `-debug=0` with `-nodebug`, drop `-debugexclude=none` check
 * Remove unused NONE enums -- fine-ish

fanquake commented:
> I'm going to open an alternate pull that does what you suggest and see what reviewers (if any) prefer.

Did this end up getting opened? Moving this PR to draft for now in any case.
